push-sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/apns/PushyApnsSender.java

```diff
@@ -197,49 +195,51 @@ private String createPushPayload(final Message message, final String pushMessage
         return payloadBuilder.buildWithDefaultMaximumLength();
     }
 
-    private ApnsClient receiveApnsConnection(final iOSVariant iOSVariant) {
+    private synchronized ApnsClient receiveApnsConnection(final iOSVariant iOSVariant) {
         return simpleApnsClientCache.getApnsClientForVariant(iOSVariant, () -> {
-            final ApnsClient apnsClient = buildApnsClient(iOSVariant);
-
-            if (apnsClient != null) {
+            final ApnsClientBuilder builder = new ApnsClientBuilder();
 
-                // connect and wait, ONLY when we have a valid client
-                logger.debug("establishing the connection for {}", iOSVariant.getVariantID());
-                connectToDestinations(iOSVariant, apnsClient);
+            assambleApnsClientBuilder(iOSVariant, builder);
+            connectToDestinations(iOSVariant, builder);
 
-                // APNS client has auto-reconnect, but let's log when that happens
-                apnsClient.getReconnectionFuture().addListener(future -> logger.trace("Reconnecting to APNs"));
+            // connect and wait, ONLY when we have a valid client
+            logger.debug("establishing the connection for {}", iOSVariant.getVariantID());
+            ApnsClient apnsClient;
+            try {
+                logger.debug("connecting to APNs");
+                apnsClient = builder.build();
                 return apnsClient;
+            } catch (SSLException e) {
+                logger.error("Error connecting to APNs", e);
             }
             return null;
+
         });
     }
 
-    private ApnsClient buildApnsClient(final iOSVariant iOSVariant) {
+    private void assambleApnsClientBuilder(final iOSVariant iOSVariant, final ApnsClientBuilder builder) {
```

Comment:
Yeah, this will be the problem... I had it in my head otherwise but then decided to not make big changes and forgot to move throw to appropriate place, I will fix it.

Reply:
yeah, previously there was a `return` statement in the body so that it will not run the `throw` line. But now since you removed the `return` line, the `throw` line is ALWAYS get executed.